src/apis/Event.re

```diff
@@ -1,66 +1,79 @@
-  "currentTarget": float,
-  "defaultPrevented": Js.Nullable.t(bool),
-  "dispatchConfig": {. "registrationName": string},
-  "eventPhase": Js.Nullable.t(float),
-  [@bs.meth] "preventDefault": unit => unit,
-  [@bs.meth] "isDefaultPrevented": unit => bool,
-  [@bs.meth] "stopPropagation": unit => unit,
-  [@bs.meth] "isPropagationStopped": unit => bool,
-  "isTrusted": Js.Nullable.t(bool),
-  "nativeEvent": 'a,
-  [@bs.meth] "persist": unit => unit,
-  "target": Js.Nullable.t(float),
-  "timeStamp": float,
-  "_type": Js.Nullable.t(string),
+type syntheticEvent('a, 'b) = {
+  bubbles: Js.Nullable.t(bool),
+  cancelable: Js.Nullable.t(bool),
+  currentTarget: float,
+  defaultPrevented: Js.Nullable.t(bool),
+  dispatchConfig: registrationName,
+  eventPhase: Js.Nullable.t(float),
+  isTrusted: Js.Nullable.t(bool),
+  nativeEvent: 'a,
+  target: Js.Nullable.t(float),
+  timeStamp: float,
+  [@bs.as "type"]
+  _type: Js.Nullable.t(string),
+  touchHistory: 'b // this field can be omitted, making 'b a phantom type
+}
+and registrationName = {registrationName: string};
+
+// type responderSyntheticEvent('a) defined as type syntheticEvent('a, touchHistory)
+
+type touchHistory = {
+  indexOfSingleActiveTouch: float,
+  mostRecentTimeStamp: float,
+  numberActiveTouches: float,
+  touchBank: array(touchBank),
+}
+and touchBank = {
+  touchActive: bool,
+  startPageX: float,
+  startPageY: float,
+  startTimeStamp: float,
+  currentPageX: float,
+  currentPageY: float,
+  currentTimeStamp: float,
+  previousPageX: float,
+  previousPageY: float,
+  previousTimeStamp: float,
 };
 
-type responderSyntheticEvent('a) = {
-  .
-  // synthethicEvent keys
-  "bubbles": Js.Nullable.t(bool),
-  "cancelable": Js.Nullable.t(bool),
-  "currentTarget": float,
-  "defaultPrevented": Js.Nullable.t(bool),
-  "dispatchConfig": {. "registrationName": string},
-  "eventPhase": Js.Nullable.t(float),
-  [@bs.meth] "preventDefault": unit => unit,
-  [@bs.meth] "isDefaultPrevented": unit => bool,
-  [@bs.meth] "stopPropagation": unit => unit,
-  [@bs.meth] "isPropagationStopped": unit => bool,
-  "isTrusted": Js.Nullable.t(bool),
-  "nativeEvent": 'a,
-  [@bs.meth] "persist": unit => unit,
-  "target": Js.Nullable.t(float),
-  "timeStamp": float,
-  "_type": Js.Nullable.t(string),
-  // responderSyntheticEvent additional key
-  "touchHistory": {
-    .
-    "indexOfSingleActiveTouch": float,
-    "mostRecentTimeStamp": float,
-    "numberActiveTouches": float,
-    "touchBank":
-      array({
-        .
-        "touchActive": bool,
-        "startPageX": float,
-        "startPageY": float,
-        "startTimeStamp": float,
-        "currentPageX": float,
-        "currentPageY": float,
-        "currentTimeStamp": float,
-        "previousPageX": float,
-        "previousPageY": float,
-        "previousTimeStamp": float,
-      }),
-  },
-};
+[@bs.get] [@bs.return nullable]
+external bubbles: syntheticEvent('a, 'b) => option(bool) = "bubbles";
```

Comment:
If I understand correctly, the benefit of this methods instead of directly hitting the record field is to have a option instead of a js nullable? I am assuming this is still zero cost (external), so how is the compiler making diff between Js.Nullable.t & option? Can't we directly type the record fields as option() & allow to hit record fields for getter & only have externals for send method? Maybe that's not optimal to have even.field & event->Event.method... ?

Reply:
So far, we have not verified if the fields may have `null` values and we have taken the safer but more annoying-to-use path to wrap everything in `Js.Nullable.t`.

`[@bs.return nullable]` provides a safe way to handle the case when the return value may be a `null`. Assume `someExternal` returns `option('a)` with the decorator and `Js.Nullable.t('a)` without.  A `switch` statement such as

```
switch (someExternal(x)) {
| None => Js.log("a")
| Some(_) => Js.log("b")
}
```
will have the below JS output when `someExternal` is defined **with** the decorator:
```
if (match == null) {
  console.log("a");
} else {
  console.log("b");
}
```
compared to the output when defined **without** the decorator
```
if (match !== undefined) {
  console.log("b");
} else {
  console.log("a");
}
```

Basically, the compiler makes use of the deep inequality operator when it needs to differentiate between `null` and `undefined` and shallow equality when it doesn't.

We have already made use of this with some of these bindings. There is no overhead for pattern matching (as in the switch statement), but `someExternal(x) == Some(y);` adds
```
var Caml_obj = require("./stdlib/caml_obj.js");
var Caml_option = require("./stdlib/caml_option.js");
```
to the JS output.

Reply:
I also would prefer to just handle actual methods using bs.send and keep everything else as record fields, even if it means that these fields may have type `Js.Nullable.t('something)`.  As a user of the API, I would find it confusing if some fields were available directly in the record and others through a getter function.